sympy/core/mul.py

```diff
@@ -1264,26 +1266,38 @@ def _eval_is_odd(self):
         is_integer = self.is_integer
 
         if is_integer:
-            r, acc = True, 1
+            #The entire expression is an integer, but it
+            #can be decomposed into fractions (i.e.
+            #numerator and denominator.
+            #The counters keep track of multiples of 2 in
+            #numerator and denominator to determine oddness.
+            count_denom = 0
+            count_num = 0
             for t in self.args:
-                if not t.is_integer:
-                    return None
-                elif t.is_even:
-                    r = False
-                elif t.is_integer:
-                    if r is False:
-                        pass
-                    elif acc != 1 and (acc + t).is_odd:
-                        r = False
-                    elif t.is_odd is None:
-                        r = None
-                acc = t
-            return r
+                if t == S.Half:
+                    count_denom += 1
+                elif isinstance(t,Symbol):
```

Comment:
Actually, I suppose more general logic could try to find the largest power of 2 that divides the denominator, and use that (I suppose you already imagined this logic with having "count_denom"). There's probably a SymPy function that returns this but I'm not sure what it is (maybe `factorint(n, limit=3).get(2, 0)`).

Also, a fact worth knowing for this logic is that a rational number can only appear once in a Mul, and it will always be the first arg. A helpful method here is `self.as_coeff_mul()`, which returns `(coeff, args)` where coeff is the numeric coefficient and `args` is the rest:

```py
>>> (2*x*y).as_coeff_mul()
(2, (x, y))
```

Also, rational numbers are always reduced to lowest terms (as you probably guessed).